File: game/src/sim/Bell.ts
// The bell: a 6-DOF rigid sphere with an internal mass-ring (a symmetric
// top). Deep-sim fidelity: full orientation + body-frame angular velocity,
// Euler's equation with the gyroscopic term, RK4. "Rings true" when spin is
// aligned with the ring axis; a clatter is an off-axis angular impulse.
//
// Body frame: ring axis = body +x̂. Inertia I = diag(Iaxial, It, It).
// A near-true spin sits on the principal axis so ω×(Iω) ≈ 0 and it *stays*
// true — physically self-consistent with the canon "rings when it spins
// true". Knock it off axis and gyroscopic precession makes it tumble.

import type { Vec3, Quat } from './vec';
import { qmul, qnorm, vnorm } from './vec';
import { accel } from './trajectory';

// Thin-ring inertia, normalised (absolute scale only sets torque response).
const I_AXIAL = 1.0; // about the ring axis (body x)
const I_TRANS = 0.5; // about a diameter (body y,z)
const SPIN_GOOD = 8; // rad/s where the chime is fully voiced
const SPIN_MIN = 1.5; // below this it can't ring at all

export interface BellBody {
  p: Vec3;
  v: Vec3;
  q: Quat; // orientation (body->world)
  w: Vec3; // angular velocity, BODY frame
}

// Rotate a body-frame vector into world using quaternion q.
function rotate(q: Quat, b: Vec3): Vec3 {
  // v' = q * (0,b) * q⁻¹  ; q is unit so q⁻¹ = conj
  const t = qmul(q, { x: b.x, y: b.y, z: b.z, w: 0 });
  const r = qmul(t, { x: -q.x, y: -q.y, z: -q.z, w: q.w });
  return { x: r.x, y: r.y, z: r.z };
}

// Euler's equation, body frame, free flight (τ_ext = 0):
//   I·ω̇ = −ω × (I·ω)
function angAccelBody(w: Vec3): Vec3 {
  const Iw = { x: I_AXIAL * w.x, y: I_TRANS * w.y, z: I_TRANS * w.z };
  const gyro = {
    x: w.y * Iw.z - w.z * Iw.y,
    y: w.z * Iw.x - w.x * Iw.z,
    z: w.x * Iw.y - w.y * Iw.x,
  };
  return { x: -gyro.x / I_AXIAL, y: -gyro.y / I_TRANS, z: -gyro.z / I_TRANS };
}

interface Deriv {
  dp: Vec3;
  dv: Vec3;
  dq: Quat;
  dw: Vec3;
}

function deriv(s: BellBody, omega: number): Deriv {
  const wWorld = rotate(s.q, s.w);
  const dqh = qmul({ x: wWorld.x, y: wWorld.y, z: wWorld.z, w: 0 }, s.q);
  return {
    dp: s.v,
    dv: accel(s.p, s.v, omega), // rotating-frame translational (sphere, no aero)
    dq: { x: 0.5 * dqh.x, y: 0.5 * dqh.y, z: 0.5 * dqh.z, w: 0.5 * dqh.w },
    dw: angAccelBody(s.w),
  };
}

function addB(s: BellBody, k: Deriv, f: number): BellBody {
  return {
    p: { x: s.p.x + k.dp.x * f, y: s.p.y + k.dp.y * f, z: s.p.z + k.dp.z * f },
    v: { x: s.v.x + k.dv.x * f, y: s.v.y + k.dv.y * f, z: s.v.z + k.dv.z * f },
    q: { x: s.q.x + k.dq.x * f, y: s.q.y + k.dq.y * f, z: s.q.z + k.dq.z * f, w: s.q.w + k.dq.w * f },
    w: { x: s.w.x + k.dw.x * f, y: s.w.y + k.dw.y * f, z: s.w.z + k.dw.z * f },
  };
}

export function stepBell(s: BellBody, omega: number, h: number): BellBody {
  const k1 = deriv(s, omega);
  const k2 = deriv(addB(s, k1, h / 2), omega);
  const k3 = deriv(addB(s, k2, h / 2), omega);
  const k4 = deriv(addB(s, k3, h), omega);
  const out: BellBody = {
    p: {
      x: s.p.x + (h / 6) * (k1.dp.x + 2 * k2.dp.x + 2 * k3.dp.x + k4.dp.x),
      y: s.p.y + (h / 6) * (k1.dp.y + 2 * k2.dp.y + 2 * k3.dp.y + k4.dp.y),
      z: s.p.z + (h / 6) * (k1.dp.z + 2 * k2.dp.z + 2 * k3.dp.z + k4.dp.z),
    },
    v: {
      x: s.v.x + (h / 6) * (k1.dv.x + 2 * k2.dv.x + 2 * k3.dv.x + k4.dv.x),
      y: s.v.y + (h / 6) * (k1.dv.y + 2 * k2.dv.y + 2 * k3.dv.y + k4.dv.y),
      z: s.v.z + (h / 6) * (k1.dv.z + 2 * k2.dv.z + 2 * k3.dv.z + k4.dv.z),
    },
    q: qnorm({
      x: s.q.x + (h / 6) * (k1.dq.x + 2 * k2.dq.x + 2 * k3.dq.x + k4.dq.x),
      y: s.q.y + (h / 6) * (k1.dq.y + 2 * k2.dq.y + 2 * k3.dq.y + k4.dq.y),
      z: s.q.z + (h / 6) * (k1.dq.z + 2 * k2.dq.z + 2 * k3.dq.z + k4.dq.z),
      w: s.q.w + (h / 6) * (k1.dq.w + 2 * k2.dq.w + 2 * k3.dq.w + k4.dq.w),
    }),
    w: {
      x: s.w.x + (h / 6) * (k1.dw.x + 2 * k2.dw.x + 2 * k3.dw.x + k4.dw.x),
      y: s.w.y + (h / 6) * (k1.dw.y + 2 * k2.dw.y + 2 * k3.dw.y + k4.dw.y),
      z: s.w.z + (h / 6) * (k1.dw.z + 2 * k2.dw.z + 2 * k3.dw.z + k4.dw.z),
    },
  };
  return out;
}

function smoothstep(a: number, b: number, x: number): number {
  const t = Math.max(0, Math.min(1, (x - a) / (b - a)));
  return t * t * (3 - 2 * t);
}

/** Spin-trueness ∈ [0,1]: alignment of ω with the body ring axis. */
export function trueness(w: Vec3): number {
  const sp = Math.sqrt(w.x * w.x + w.y * w.y + w.z * w.z);
  if (sp < 1e-6) return 0;
  const align = Math.abs(w.x) / sp; // ring axis = body x
  return Math.max(0, Math.min(1, align));
}

/** Audible chime ∈ [0,1]: trueness gated by enough spin to voice. */
export function chime(w: Vec3): number {
  const sp = Math.sqrt(w.x * w.x + w.y * w.y + w.z * w.z);
  return trueness(w) * smoothstep(SPIN_MIN, SPIN_GOOD, sp);
}

/** A defender knocks it off true: add an off-axis body-frame angular impulse. */
export function clatter(s: BellBody, bodyImpulse: Vec3): void {
  s.w = {
    x: s.w.x + bodyImpulse.x / I_AXIAL,
    y: s.w.y + bodyImpulse.y / I_TRANS,
    z: s.w.z + bodyImpulse.z / I_TRANS,
  };
}

export { rotate as rotateBodyToWorld, vnorm };

File: game/rig-core/src/bell.rs
//! 6-DOF bell with internal mass-ring (symmetric top) — port of TS sim/Bell.ts.
//! Body frame: ring axis = body +x̂. Inertia I = diag(I_AXIAL, I_TRANS, I_TRANS).
//! A true spin sits on the principal axis (ω×(Iω) ≈ 0) and stays true;
//! an off-axis impulse induces gyroscopic precession and tumble.

use crate::math::{Quat, Vec3};
use crate::trajectory::accel;

// Thin-ring inertia, normalised.
const I_AXIAL: f64 = 1.0; // about the ring axis (body x)
const I_TRANS: f64 = 0.5; // about a diameter (body y,z)
const SPIN_GOOD: f64 = 8.0; // rad/s where the chime is fully voiced
const SPIN_MIN: f64 = 1.5; // below this it cannot ring at all

/// Full rigid-body state: centre-of-mass position/velocity + orientation/spin.
#[derive(Clone, Copy, Debug)]
pub struct BellBody {
    pub p: Vec3,
    pub v: Vec3,
    pub q: Quat, // orientation body → world (unit quaternion)
    pub w: Vec3, // angular velocity in BODY frame (rad/s)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/// Euler's equation, body frame, free flight (τ_ext = 0):
///   I·ω̇ = −ω × (I·ω)
fn ang_accel_body(w: Vec3) -> Vec3 {
    let iw = Vec3::new(I_AXIAL * w.x, I_TRANS * w.y, I_TRANS * w.z);
    let gyro = w.cross(iw);
    Vec3::new(-gyro.x / I_AXIAL, -gyro.y / I_TRANS, -gyro.z / I_TRANS)
}

/// All four time-derivatives for (p, v, q, w).
fn deriv_bell(s: BellBody, omega: f64) -> (Vec3, Vec3, Quat, Vec3) {
    let w_world = s.q.rotate(s.w);
    let dq = s.q.deriv(w_world);
    (s.v, accel(s.p, s.v, omega), dq, ang_accel_body(s.w))
}

/// Linear blend:  s + k * f  (componentwise over all four fields).
fn add_scaled(s: BellBody, dp: Vec3, dv: Vec3, dq: Quat, dw: Vec3, f: f64) -> BellBody {
    BellBody {
        p: s.p.add(dp.scale(f)),
        v: s.v.add(dv.scale(f)),
        q: Quat::new(
            s.q.x + dq.x * f,
            s.q.y + dq.y * f,
            s.q.z + dq.z * f,
            s.q.w + dq.w * f,
        ),
        w: s.w.add(dw.scale(f)),
    }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Advance the bell by one RK4 step of size `h` seconds.
/// The quaternion is renormalised at the end of every step.
pub fn step_bell(s: BellBody, omega: f64, h: f64) -> BellBody {
    let (k1p, k1v, k1q, k1w) = deriv_bell(s, omega);
    let s2 = add_scaled(s, k1p, k1v, k1q, k1w, h / 2.0);

    let (k2p, k2v, k2q, k2w) = deriv_bell(s2, omega);
    let s3 = add_scaled(s, k2p, k2v, k2q, k2w, h / 2.0);

    let (k3p, k3v, k3q, k3w) = deriv_bell(s3, omega);
    let s4 = add_scaled(s, k3p, k3v, k3q, k3w, h);

    let (k4p, k4v, k4q, k4w) = deriv_bell(s4, omega);

    let c = h / 6.0;
    BellBody {
        p: Vec3::new(
            s.p.x + c * (k1p.x + 2.0 * k2p.x + 2.0 * k3p.x + k4p.x),
            s.p.y + c * (k1p.y + 2.0 * k2p.y + 2.0 * k3p.y + k4p.y),
            s.p.z + c * (k1p.z + 2.0 * k2p.z + 2.0 * k3p.z + k4p.z),
        ),
        v: Vec3::new(
            s.v.x + c * (k1v.x + 2.0 * k2v.x + 2.0 * k3v.x + k4v.x),
            s.v.y + c * (k1v.y + 2.0 * k2v.y + 2.0 * k3v.y + k4v.y),
            s.v.z + c * (k1v.z + 2.0 * k2v.z + 2.0 * k3v.z + k4v.z),
        ),
        q: Quat::new(
            s.q.x + c * (k1q.x + 2.0 * k2q.x + 2.0 * k3q.x + k4q.x),
            s.q.y + c * (k1q.y + 2.0 * k2q.y + 2.0 * k3q.y + k4q.y),
            s.q.z + c * (k1q.z + 2.0 * k2q.z + 2.0 * k3q.z + k4q.z),
            s.q.w + c * (k1q.w + 2.0 * k2q.w + 2.0 * k3q.w + k4q.w),
        )
        .norm(),
        w: Vec3::new(
            s.w.x + c * (k1w.x + 2.0 * k2w.x + 2.0 * k3w.x + k4w.x),
            s.w.y + c * (k1w.y + 2.0 * k2w.y + 2.0 * k3w.y + k4w.y),
            s.w.z + c * (k1w.z + 2.0 * k2w.z + 2.0 * k3w.z + k4w.z),
        ),
    }
}

/// Spin-trueness ∈ [0,1]: alignment of ω with the body ring axis (body x̂).
/// Returns 0 if |ω| < 1e-6.
pub fn trueness(w: Vec3) -> f64 {
    let sp = w.len();
    if sp < 1e-6 {
        return 0.0;
    }
    (w.x.abs() / sp).clamp(0.0, 1.0)
}

/// Audible chime ∈ [0,1]: trueness gated by enough spin to voice.
pub fn chime(w: Vec3) -> f64 {
    let sp = w.len();
    trueness(w) * smoothstep(SPIN_MIN, SPIN_GOOD, sp)
}

/// Apply an off-axis body-frame angular impulse (e.g. a defender's knock).
/// Each component is divided by the corresponding principal inertia.
pub fn clatter(s: &mut BellBody, body_impulse: Vec3) {
    s.w.x += body_impulse.x / I_AXIAL;
    s.w.y += body_impulse.y / I_TRANS;
    s.w.z += body_impulse.z / I_TRANS;
}

// ---------------------------------------------------------------------------
// Private utility
// ---------------------------------------------------------------------------

fn smoothstep(a: f64, b: f64, x: f64) -> f64 {
    let t = ((x - a) / (b - a)).clamp(0.0, 1.0);
    t * t * (3.0 - 2.0 * t)
}

// ---------------------------------------------------------------------------
// Tests — mirror of test/bell.test.ts
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::trajectory::analytic;

    const W: f64 = 0.32;
    const H: f64 = 1.0 / 240.0;

    fn mk(w: Vec3) -> BellBody {
        BellBody {
            p: Vec3::new(0.0, 4.0, -2.0),
            v: Vec3::new(18.0, -3.0, 5.0),
            q: Quat::ident(),
            w,
        }
    }

    fn default_bell() -> BellBody {
        mk(Vec3::new(30.0, 0.0, 0.0))
    }

    #[test]
    fn com_follows_analytic_coriolis_path() {
        let mut s = default_bell();
        let p0 = s.p;
        let v0 = s.v;
        for i in 1..=(240 * 5) {
            s = step_bell(s, W, H);
            let t = i as f64 * H;
            let a = analytic(p0, v0, W, t);
            assert!((s.p.y - a.y).abs() < 1e-2, "y mismatch at t={t}: {} vs {}", s.p.y, a.y);
            assert!((s.p.z - a.z).abs() < 1e-2, "z mismatch at t={t}: {} vs {}", s.p.z, a.z);
            assert!(
                (s.p.x - v0.x * t).abs() < 1e-4,
                "x axial mismatch at t={t}: {} vs {}",
                s.p.x,
                v0.x * t
            );
        }
    }

    #[test]
    fn true_spin_stays_true() {
        let mut s = mk(Vec3::new(28.0, 0.0, 0.0));
        for _ in 0..(240 * 4) {
            s = step_bell(s, W, H);
        }
        assert!(trueness(s.w) > 0.999, "trueness = {}", trueness(s.w));
        assert!(chime(s.w) > 0.9, "chime = {}", chime(s.w));
    }

    #[test]
    fn off_axis_spin_tumbles_and_precesses() {
        let mut s = mk(Vec3::new(14.0, 9.0, 6.0));
        let t0 = trueness(s.w);
        let mut moved = false;
        let mut prev_wy = s.w.y;
        for _ in 0..(240 * 3) {
            s = step_bell(s, W, H);
            if (s.w.y - prev_wy).abs() > 1e-3 {
                moved = true;
            }
            prev_wy = s.w.y;
        }
        assert!(t0 < 0.85, "initial trueness should be < 0.85, got {t0}");
        assert!(chime(s.w) < 0.9, "chime should drop, got {}", chime(s.w));
        assert!(moved, "gyroscopic precession should move ω_y");
    }

    #[test]
    fn angular_momentum_magnitude_conserved() {
        let lm = |b: &BellBody| -> f64 {
            // |I·ω| as a body-frame proxy
            (I_AXIAL * b.w.x).hypot((I_TRANS * b.w.y).hypot(I_TRANS * b.w.z))
        };
        let mut s = mk(Vec3::new(14.0, 9.0, 6.0));
        let l0 = lm(&s);
        for _ in 0..(240 * 3) {
            s = step_bell(s, W, H);
        }
        let l1 = lm(&s);
        assert!((l1 - l0).abs() < 0.1, "|I·ω| drift: {l0} → {l1}");
    }

    #[test]
    fn clatter_drops_trueness() {
        let mut s = mk(Vec3::new(28.0, 0.0, 0.0));
        assert!(trueness(s.w) > 0.999);
        clatter(&mut s, Vec3::new(0.0, 12.0, 8.0));
        assert!(trueness(s.w) < 0.92, "trueness after clatter = {}", trueness(s.w));
    }
}
